File: node/src/progress.ts
import ora from "ora";

export type SyncAction = "remove" | "install" | "update";

export interface SyncProgressReporter {
  begin(total: number, message?: string): void;
  removing(name: string, index: number, total: number): void;
  downloading(name: string, relPath: string, fileIndex: number, fileTotal: number): void;
  installing(name: string): void;
  linking(name: string): void;
}

export class OraSyncProgress implements SyncProgressReporter {
  private spinner = ora("Applying changes…").start();
  private total = 1;
  private completed = 0;

  begin(total: number, message = "Applying changes"): void {
    this.total = Math.max(total, 1);
    this.completed = 0;
    this.spinner.text = `${message} (0/${this.total})`;
  }

  private pct(): string {
    const percent = Math.min(100, Math.round((this.completed / this.total) * 100));
    return `${percent}%`;
  }

  removing(name: string, index: number, total: number): void {
    this.completed += 1;
    this.spinner.text = `[${this.pct()}] Removing ${name} (${index}/${total})`;
  }

  downloading(name: string, relPath: string, fileIndex: number, fileTotal: number): void {
    this.spinner.text = `[${this.pct()}] Downloading ${name} — ${relPath} (${fileIndex}/${fileTotal})`;
  }

  installing(name: string): void {
    this.completed += 1;
    this.spinner.text = `[${this.pct()}] Installing ${name}`;
  }

  linking(name: string): void {
    this.spinner.text = `[${this.pct()}] Linking ${name}`;
  }

  succeed(message: string): void {
    this.spinner.succeed(message);
  }

  fail(message: string): void {
    this.spinner.fail(message);
  }
}

File: python/src/skills_utils/progress.py
from __future__ import annotations

from rich.progress import BarColumn, Progress, SpinnerColumn, TextColumn, TimeElapsedColumn


class SyncProgress:
    """Rich progress display for skill sync operations."""

    def __init__(self) -> None:
        self._progress = Progress(
            SpinnerColumn(),
            TextColumn("[bold blue]{task.description}"),
            BarColumn(bar_width=32),
            TextColumn("[progress.percentage]{task.percentage:>3.0f}%"),
            TimeElapsedColumn(),
            transient=False,
        )
        self._task_id: int | None = None
        self._total = 0

    def __enter__(self) -> SyncProgress:
        self._progress.start()
        return self

    def __exit__(self, *args: object) -> None:
        self._progress.stop()

    def begin(self, total: int, message: str = "Applying changes") -> None:
        self._total = max(total, 1)
        self._task_id = self._progress.add_task(message, total=self._total)

    def _update(self, description: str, advance: int = 0) -> None:
        if self._task_id is None:
            return
        self._progress.update(self._task_id, description=description, advance=advance)

    def removing(self, name: str, index: int, total: int) -> None:
        self._update(f"Removing {name} ({index}/{total})", advance=1)

    def downloading(self, name: str, rel_path: str, file_index: int, file_total: int) -> None:
        self._update(f"Downloading {name} — {rel_path} ({file_index}/{file_total})")

    def installing(self, name: str) -> None:
        self._update(f"Installing {name}", advance=1)

    def linking(self, name: str) -> None:
        self._update(f"Linking {name} for agent discovery")
